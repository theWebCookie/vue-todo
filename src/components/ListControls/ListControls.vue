<template>
  <div :class="$style.controls">
    <div>
      <span v-if="tasksLeft === 0">No tasks left</span>
      <span v-else-if="tasksLeft === 1">1 task left</span>
      <span v-else>{{ tasksLeft }} tasks left</span>
    </div>
    <div :class="$style.filterButtons">
      <button @click="filterTasks('all')">All</button>
      <button @click="filterTasks('active')">Active</button>
      <button @click="filterTasks('completed')">Completed</button>
    </div>
    <div :class="$style.clearButton">
      <button @click="clearCompleted">Clear Completed</button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'ListControls',
  computed: {
    tasksLeft() {
      return this.$store.getters.tasksLeft;
    },
  },
  methods: {
    filterTasks(filter) {
      this.$store.dispatch('setFilter', filter);
    },
    clearCompleted() {
      this.$store.dispatch('clearCompletedTasks');
    },
  },
};
</script>

<style module lang="scss">
.controls {
  display: flex;
  align-items: center;
  padding: 10px;
  background-color: #fff;
  font-size: 14px;
  justify-content: space-between;
  color: #b0b0b0;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;

  .filterButtons button,
  .clearButton button {
    all: unset;
    cursor: pointer;
    transition: 0.3s;

    &:hover {
      color: #007bff;
    }
  }
  .filterButtons button {
    margin: 0 5px;
  }
}
</style>
