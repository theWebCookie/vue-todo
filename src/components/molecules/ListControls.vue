<template>
  <div :class="$style.controls">
    <div>
      <span v-if="tasksLeft === 0">No tasks left</span>
      <span v-else-if="tasksLeft === 1">1 task left</span>
      <span v-else>{{ tasksLeft }} tasks left</span>
    </div>
    <div :class="$style.filterButtons">
      <AppButton v-for="filter in filters" :key="filter.name" :customClass="$style.filterButton" @click="filterTasks(filter.value)">
        {{ filter.name }}
      </AppButton>
    </div>
    <div :class="$style.clearButton">
      <AppButton @click="clearCompleted" :customClass="$style.clearButtonStyle"> Clear Completed </AppButton>
    </div>
  </div>
</template>

<script>
import AppButton from '../atoms/AppButton.vue';

export default {
  name: 'ListControls',
  components: {
    AppButton,
  },
  computed: {
    tasksLeft() {
      return this.$store.getters.tasksLeft;
    },
    filters() {
      return [
        { name: 'All', value: 'all' },
        { name: 'Active', value: 'active' },
        { name: 'Completed', value: 'completed' },
      ];
    },
  },
  methods: {
    filterTasks(filter) {
      this.$store.dispatch('setFilter', filter);
    },
    clearCompleted() {
      this.$store.dispatch('clearCompletedTasks');
    },
  },
};
</script>

<style module lang="scss">
.controls {
  display: flex;
  align-items: center;
  padding: 10px;
  background-color: #fff;
  font-size: 14px;
  justify-content: space-between;
  color: #b0b0b0;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;

  .filterButtons {
    display: flex;
  }

  .filterButton {
    margin: 0 5px;
  }
}
</style>
