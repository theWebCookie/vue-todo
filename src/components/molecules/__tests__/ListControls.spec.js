import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from '@jest/globals';
import { createStore } from 'vuex';
import ListControls from '../ListControls.vue';

const styles = {
  controls: 'controls',
  filterButtons: 'filterButtons',
  filterButton: 'filterButton',
  clearButton: 'clearButton',
  clearButtonStyle: 'clearButtonStyle',
};

const numberOfTasksLeft = 5;

describe('ListControls.vue', () => {
  let store;
  let getters;

  beforeEach(() => {
    getters = {
      tasksLeft: () => numberOfTasksLeft,
    };

    store = createStore({
      getters,
    });
  });

  it('renders correctly and matches the snapshot', () => {
    const wrapper = shallowMount(ListControls, {
      global: {
        plugins: [store],
        mocks: {
          $style: styles,
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct task count message', () => {
    const wrapper = shallowMount(ListControls, {
      global: {
        plugins: [store],
        mocks: {
          $style: styles,
        },
      },
    });

    const taskCountMessage = wrapper.find('span').text();
    expect(taskCountMessage).toBe(`${numberOfTasksLeft} tasks left`);
  });
});
