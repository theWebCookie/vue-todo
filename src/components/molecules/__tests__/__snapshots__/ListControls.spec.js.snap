// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListControls.vue renders correctly and matches the snapshot 1`] = `
<div
  class="controls"
>
  <div>
    <span>
      5 tasks left
    </span>
  </div>
  <div
    class="filterButtons"
  >
    
    <app-button-stub
      customclass="filterButton"
      isactive="false"
    />
    <app-button-stub
      customclass="filterButton"
      isactive="false"
    />
    <app-button-stub
      customclass="filterButton"
      isactive="false"
    />
    
  </div>
  <div
    class=""
  >
    <app-button-stub
      customclass="clearButtonStyle"
      disabled="false"
      isactive="false"
    />
  </div>
</div>
`;
